src/services/attachmentService.js

```diff
@@ -0,0 +1,96 @@
+import axios from '@nextcloud/axios'
+import { generateOcsUrl } from '@nextcloud/router'
+import { showError, showSuccess } from '@nextcloud/dialogs'
+import { translate as t } from '@nextcloud/l10n'
+import { async } from 'regenerator-runtime'
+
+/**
+ * Makes a share link for a given file or directory.
+ * @param {string} path The file path from the user's root directory. e.g. `/myfile.txt`
+ * @param {string} token The conversation's token
+ * @returns {string} url share link
+ */
+const shareFile = async function(path, token) {
+	try {
+		const res = await axios.post(generateOcsUrl('apps/files_sharing/api/v1/', 2) + 'shares', {
+			shareType: 3, // OC.Share.SHARE_TYPE_LINK,
+			path,
+			shareWith: token,
+		})
+		return res.data.ocs.data
+	} catch (error) {
+		if (
+			error.response
+			&& error.response.data
+			&& error.response.data.ocs
+			&& error.response.data.ocs.meta
+			&& error.response.data.ocs.meta.message
+		) {
+			console.error(`Error while sharing file: ${error.response.data.ocs.meta.message || 'Unknown error'}`)
+			showError(error.response.data.ocs.meta.message)
+			throw error
+		} else {
+			console.error('Error while sharing file: Unknown error')
+			showError(t('mail', 'Error while sharing file'))
+			throw error
+		}
+	}
+}
+
+const shareFileWith = async function(path, token, sharedWith, permissions = 17){
+	try {
+		const res = await axios.post(generateOcsUrl('apps/files_sharing/api/v1/', 2) + 'shares', {
+			password: null,
+			shareType: 0, // WITH USERS,
+			permissions: permissions, // 14 - edit, 17 - view
+			path,
+			shareWith: sharedWith,
+		})
+		return res.data.ocs.data
+	} catch (error) {
+		console.error(error)
+	}
+}
+
+const uploadLocalAttachment = async function(event, dav, componentAttachments) {
+	const files = event.target.files
+	let attachments = [];
+	let promises = [];
+	
+	files.forEach(file => {
+		if(componentAttachments.map(attachment => attachment.fileName).indexOf(file.name) !== -1){
+			showError(t('calendar', 'Attachment {fileName} already exists!',{
+				fileName: file.name
+			}))
+		}
+		else {
+			const url = `/remote.php/dav/files/${dav.userId}/${file.name}`
```

Comment:
There's the brand new `nextcloud-upload` package from @skjnldsv that might be a better fit: https://github.com/nextcloud/nextcloud-upload.

Reply:
> @sazanof Tests have to be updated for the new `attachments` property.

Tell me, is it enough to add a parameter or do I need to write a full test? If it's a complete test, then I ask for help, since **I've never written tests and don't understand how to write them** (It's a matter of time and effort, I don't argue). =( 
I tried to write my own block for checking (by analogy with other tests) `it()`
This error occurs if I add my parameters to 
```
expect(getDefaultEvent Object()).to Equal({
....
attachments: [
	'ATTACHMENT1',
	'ATTACHMENT2',
	'ATTACHMENT3',
	'ATTACHMENT4',
	]
})
```
https://gist.github.com/sazanof/ecb2aaa8c760f93e74d2952294ba09ba

When checking, I get an error. I can't understand why it occurs...
```
 ● Test suite failed to run

    TypeError: Converting circular structure to JSON
        --> --> --> --> --> --> --> --> starting at object with constructor 'CalendarComponent'
        --- property '_root' closes the circle
        at stringify (<anonymous>)

```

Reply:
Yes, clearly.

Reply:
@skjnldsv **Well, since we are not in a hurry, we can discuss =).** 

I would like to be able to upload to a specific folder + have a callback to provide shared access or something else. Maybe a subscription to some events (loading before, progress, after) ...
By buttons - you need a universal button that can be entered into any standard NC component. Ex.: ListItem,ActionButton, Button and so on. 

It would also be cool to be able to set the template of the uploaded file (before uploading, at the moment and after).

If you make the component universal, then in the future consider using it to upload not only to DAV, but also when writing a letter in the mail, for example...